shell/platform/darwin/macos/framework/Source/FlutterResizeSynchronizer.mm

```diff
@@ -0,0 +1,97 @@
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterResizeSynchronizer.h"
+
+#import <mutex>
+
+@interface FlutterResizeSynchronizer () {
+  uint32_t cookie;  // counter to detect stale callbacks
+  std::mutex mutex;
+  std::condition_variable condRun;
+  std::condition_variable condBlock;
+  bool accepting;
+  bool waiting;
+  bool pendingCommit;
+  CGSize newSize;
+  __weak id<FlutterResizeSynchronizerDelegate> delegate;
+}
+@end
+
+@implementation FlutterResizeSynchronizer
+
+- (instancetype)initWithDelegate:(id<FlutterResizeSynchronizerDelegate>)delegate_ {
+  if (self = [super init]) {
+    accepting = true;
+    delegate = delegate_;
+  }
+  return self;
+}
+
+- (void)beginResize:(CGSize)size notify:(dispatch_block_t)notify {
+  std::unique_lock<std::mutex> lock(mutex);
+  if (!delegate) {
+    return;
+  }
+
+  ++cookie;
+
+  // from now on, ignore all incoming commits until the block below gets
+  // scheduled on raster thread
+  accepting = false;
+
+  // let pending commits finish to unblock the raster thread
+  condRun.notify_all();
+
+  // let the engine send resize notification
+  notify();
+
+  newSize = size;
+
+  waiting = true;
+
+  condBlock.wait(lock);
```

Comment:
Better to [wait with a condition](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-wait) here. Maybe the condition here is `[]() { return pendingCommit; }`.

Reply:
Lets do it for all `wait` calls in this class.

Reply:
The version before that was running the thread loop didn't have this problem. Definitely need to check for spurious wake-up.

Reply:
https://github.com/flutter/engine/pull/21525/commits/d046e86ca0630a1edb433cd7d46306a52e7ee111